File: cpp/SegmentTree.cpp
template<typename T,  T(*op)(T, T)>
struct SegmentTree {
    // 1-index tree
    private:
    int n = 1;
    vector<T> node;
    T e;

    public:
    SegmentTree(int _n, T e): e(e){
        while(n < _n)n <<= 1;
        node.resize(2*n, e);
    }

    SegmentTree(const vector<T>& elem, T e):e(e){
        int _n = elem.size();
        while(n < _n) n <<= 1;
        node.resize(2*n, e);

        for(int i = 0;i < _n; ++i){
            set(i, elem[i]);
        }
        build();
    }

    void set(int i, T val) {
        assert(0 <= i && i < n);
        node[i + n] = val;
    }

    void build() {
        for(int k = n-1; k > 0; --k) {
            node[k] = op(node[k<<1], node[k << 1 | 1]);
        }
    }

    void update(int x, T val){
        assert(0 <= x && x < n);
        x += n;

        node[x] = val;
        while(x > 1){
            x >>= 1;
            node[x] = op(node[x<<1], node[x<<1 | 1]);
        }
    }

    //[l, r)
    T query(int l, int r)const {
        assert(0 <= l && l <= n);
        assert(0 <= r && r <= n);
        l += n;
        r += n;
        T ans_l = e;
        T ans_r = e;
        while(l < r){
            if(l&1){
                ans_l = op(ans_l, node[l]);
                ++l;
            }
            if(r&1){
                --r;
                ans_r = op(node[r], ans_r);
            }
            l >>= 1;
            r >>= 1;
        }

        return op(ans_l, ans_r);
    }

    T get(int i) const{
        assert(0 <= i && i < n);
        i += n;
        return node[i];
    }
};

File: rust/SegmentTree.rs
pub trait Monoid {
    fn e() -> Self;

    fn op(&self, rhs: &Self) -> Self;
}

pub struct  SegmentTree<M> {
    n: usize,
    node: Vec<M>,
}

impl<M: Monoid + Clone> From<&Vec<M>> for SegmentTree<M> {
    fn from(value: &Vec<M>) -> Self {
        let n = value.len();
        let mut tree = SegmentTree::new(n);
        for i in 0..n {
            tree.set(i, value[i].clone());
        }

        tree
    }
}

impl<M: Monoid> SegmentTree<M> {
    pub fn new(mut n: usize) -> SegmentTree<M> {
        n = n.next_power_of_two();
        let node = (0..2*n).map(|_| M::e()).collect();
        SegmentTree{
            n,
            node,
        }
    }

    pub fn set(&mut self, mut i: usize, val: M) {
        assert!(i < self.n);
        i += self.n;

        self.node[i] = val;
        while i > 1 {
            i >>= 1;
            self.node[i] = self.node[i<<1].op(&self.node[i<<1 | 1]);
        }
    }

    pub fn query<R>(&self, rng: R) -> M
    where R: std::ops::RangeBounds<usize>
    {
        let mut l = match rng.start_bound() {
            std::ops::Bound::Excluded(&v) => v+1,
            std::ops::Bound::Included(&v) => v,
            std::ops::Bound::Unbounded => 0,
        };
        let mut r = match rng.end_bound() {
            std::ops::Bound::Excluded(&v) => v,
            std::ops::Bound::Included(&v) => v+1,
            std::ops::Bound::Unbounded => self.n,
        };
        assert!(l <= r && r <= self.n);

        l += self.n;
        r += self.n;

        let mut ans_l = M::e();
        let mut ans_r = M::e();
        while l < r {
            if l&1 == 1 {
                ans_l = ans_l.op(&self.node[l]);
                l += 1;
            }
            if r&1 == 1 {
                r -= 1;
                ans_r = self.node[r].op(&ans_r)
            }
            l >>= 1;
            r >>= 1;
        }

        ans_l.op(&ans_r)
    }

    pub fn get(&self, i: usize) -> &M {
        assert!(i < self.n);
        &self.node[i + self.n]
    }
}
